Assignment 1
COMP 499

<p style='text-align: right;'> Data Analysis and Visualization in Python for Ecologists </p>

# Data Types and Formats

## Types of Data

The format of individual columns and rows will impact analysis performed on a dataset read into python. For example, you can’t perform mathematical calculations on a string (text formatted data). This might seem obvious, however sometimes numeric values are read into Python as strings. In this situation, when you then try to perform calculations on the string-formatted numeric data, you get an error.

In this lesson we will review ways to explore and better understand the structure and format of our data.

### Numeric Data Types

Numeric data types include integers and floats. A floating point (known as a float) number has decimal points even if that decimal point value is 0. For example: 1.13, 2.0, 1234.345. If we have a column that contains both integers and floating point numbers, Pandas will assign the entire column to the float data type so the decimal points are not lost.

An integer will never have a decimal point. Thus if we wanted to store 1.13 as an integer it would be stored as 1. Similarly, 1234.345 would be stored as 1234. You will often see the data type Int64 in Python which stands for 64 bit integer. The 64 simply refers to the memory allocated to store data in each cell which effectively relates to how many digits it can store in each “cell”. Allocating space ahead of time allows computers to optimize storage and processing efficiency.

### Text Data Type

Text data type is known as Strings in Python, or Objects in Pandas. Strings can contain numbers and / or characters. For example, a string might be a word, a sentence, or several sentences. A Pandas object might also be a plot name like ‘plot1’. A string can also contain or consist of numbers. For instance, ‘1234’ could be stored as a string. As could ‘10.23’. However strings that contain numbers can not be used for mathematical operations!

Pandas and base Python use slightly different names for data types. More on this is in the table below:

<table>
  <thead>
    <tr>
      <th>Pandas Type</th>
      <th>Native Python Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>object</td>
      <td>string</td>
      <td>The most general dtype. Will be assigned to your column if column has mixed types (numbers and strings).</td>
    </tr>
    <tr>
      <td>int64</td>
      <td>int</td>
      <td>Numeric characters. 64 refers to the memory allocated to hold this character.</td>
    </tr>
    <tr>
      <td>float64</td>
      <td>float</td>
      <td>Numeric characters with decimals. If a column contains numbers and NaNs (see below), pandas will default to float64, in case your missing value has a decimal.</td>
    </tr>
    <tr>
      <td>datetime64, timedelta[ns]</td>
      <td>N/A (but see the <a href="http://doc.python.org/2/library/datetime.html">datetime</a> module in Python’s standard library)</td>
      <td>Values meant to hold time data. Look into these for time series experiments.</td>
    </tr>
  </tbody>
</table>


## Checking the format of our data

Now that we’re armed with a basic understanding of numeric and text data types, let’s explore the format of our survey data. We’ll be working with the same surveys.csv dataset that we’ve used in previous lessons.

In [6]:
import pandas as pd
surveys_df = pd.read_csv("data/surveys.csv")

Remember that we can check the type of an object like this:

In [7]:
type(surveys_df)

pandas.core.frame.DataFrame

Next, let’s look at the structure of our surveys data. In pandas, we can check the type of one column in a DataFrame using the syntax dataFrameName[column_name].dtype:

In [8]:
surveys_df['sex'].dtype

dtype('O')

A type ‘O’ just stands for “object” which in Pandas’ world is a string (text).

In [9]:
surveys_df['record_id'].dtype

dtype('int64')

The type int64 tells us that Python is storing each value within this column as a 64 bit integer. We can use the dat.dtypes command to view the data type for each column in a DataFrame (all at once).

In [10]:
surveys_df.dtypes

record_id            int64
month                int64
day                  int64
year                 int64
plot_id              int64
species_id          object
sex                 object
hindfoot_length    float64
weight             float64
dtype: object

Note that most of the columns in our Survey data are of type int64. This means that they are 64 bit integers. But the weight column is a floating point value which means it contains decimals. The species_id and sex columns are objects which means they contain strings.



## Working With Integers and Floats

So we’ve learned that computers store numbers in one of two ways: as integers or as floating-point numbers (or floats). Integers are the numbers we usually count with. Floats have fractional parts (decimal places). Let’s next consider how the data type can impact mathematical operations on our data. Addition, subtraction, division and multiplication work on floats and integers as we’d expect.

In [11]:
print(5+5)
10

print(24-4)
20

10
20


20

If we divide one integer by another, we get a float. The result on Python 3 is different than in Python 2, where the result is an integer (integer division).

In [18]:
print(5/9)

print(10/3)

0.5555555555555556
3.3333333333333335


We can also convert a floating point number to an integer or an integer to floating point number. Notice that Python by default rounds down when it converts from floating point to integer.

In [16]:
# Convert a to an integer
a = 7.83
int(a)

7

In [17]:
# Convert b to a float
b = 7
float(b)

7.0

## Working With Our Survey Data

Getting back to our data, we can modify the format of values within our data, if we want. For instance, we could convert the record_id field to floating point values.

In [33]:
# Convert the record_id field from an integer to a float
surveys_df['record_id'] = surveys_df['record_id'].astype('float64')
    # astype: Copy of the array, cast to a specified type.
surveys_df['record_id'].dtype

dtype('float64')

### Challenge - Changing Types

Try converting the column plot_id to floats using

In [35]:
surveys_df.plot_id.astype("float")

0         2.0
1         3.0
2         2.0
3         7.0
4         3.0
5         1.0
6         2.0
7         1.0
8         1.0
9         6.0
10        5.0
11        7.0
12        3.0
13        8.0
14        6.0
15        4.0
16        3.0
17        2.0
18        4.0
19       11.0
20       14.0
21       15.0
22       13.0
23       13.0
24        9.0
25       15.0
26       15.0
27       11.0
28       11.0
29       10.0
         ... 
35519     9.0
35520     9.0
35521     9.0
35522     9.0
35523     9.0
35524     9.0
35525     8.0
35526    13.0
35527    13.0
35528    13.0
35529    13.0
35530    13.0
35531    14.0
35532    14.0
35533    14.0
35534    14.0
35535    14.0
35536    14.0
35537    15.0
35538    15.0
35539    15.0
35540    15.0
35541    15.0
35542    15.0
35543    15.0
35544    15.0
35545    15.0
35546    10.0
35547     7.0
35548     5.0
Name: plot_id, Length: 35549, dtype: float64

Next try converting weight to an integer. What goes wrong here? What is Pandas telling you? We will talk about some solutions to this later.

In [38]:
surveys_df['weight'].astype("int")

ValueError: Cannot convert non-finite values (NA or inf) to integer

## Missing Data Values - NaN

What happened in the last challenge activity? Notice that this throws a value error: ValueError: Cannot convert NA to integer. If we look at the weight column in the surveys data we notice that there are NaN (Not a Number) values. NaN values are undefined values that cannot be represented mathematically. Pandas, for example, will read an empty cell in a CSV or Excel sheet as a NaN. NaNs have some desirable properties: if we were to average the weight column without replacing our NaNs, Python would know to skip over those cells.

In [34]:
surveys_df['weight'].mean()

42.672428212991356

Dealing with missing data values is always a challenge. It’s sometimes hard to know why values are missing - was it because of a data entry error? Or data that someone was unable to collect? Should the value be 0? We need to know how missing values are represented in the dataset in order to make good decisions. If we’re lucky, we have some metadata that will tell us more about how null values were handled.

For instance, in some disciplines, like Remote Sensing, missing data values are often defined as -9999. Having a bunch of -9999 values in your data could really alter numeric calculations. Often in spreadsheets, cells are left empty where no data are available. Pandas will, by default, replace those missing values with NaN. However it is good practice to get in the habit of intentionally marking cells that have no data, with a no data value! That way there are no questions in the future when you (or someone else) explores your data.

### Where Are the NaN’s?

Let’s explore the NaN values in our data a bit further. Using the tools we learned in lesson 02, we can figure out how many rows contain NaN values for weight. We can also create a new subset from our data that only contains rows with weight values > 0 (i.e., select meaningful weight values):

In [39]:
len(surveys_df[pd.isnull(surveys_df.weight)])

3266

In [40]:
# How many rows have weight values?
len(surveys_df[surveys_df.weight> 0])

32283

We can replace all NaN values with zeroes using the .fillna() method (after making a copy of the data so we don’t lose our work):

In [41]:
df1 = surveys_df.copy()
# Fill all NaN values with 0
df1['weight'] = df1['weight'].fillna(0)

However NaN and 0 yield different analysis results. The mean value when NaN values are replaced with 0 is different from when NaN values are simply thrown out or ignored.

In [42]:
df1['weight'].mean()

38.751976145601844

We can fill NaN values with any value that we chose. The code below fills all NaN values with a mean for all weight values.

In [43]:
df1['weight'] = surveys_df['weight'].fillna(surveys_df['weight'].mean())

In [44]:
print(df1['weight'])

0        42.672428
1        42.672428
2        42.672428
3        42.672428
4        42.672428
5        42.672428
6        42.672428
7        42.672428
8        42.672428
9        42.672428
10       42.672428
11       42.672428
12       42.672428
13       42.672428
14       42.672428
15       42.672428
16       42.672428
17       42.672428
18       42.672428
19       42.672428
20       42.672428
21       42.672428
22       42.672428
23       42.672428
24       42.672428
25       42.672428
26       42.672428
27       42.672428
28       42.672428
29       42.672428
           ...    
35519    36.000000
35520    48.000000
35521    45.000000
35522    44.000000
35523    27.000000
35524    26.000000
35525    24.000000
35526    43.000000
35527    42.672428
35528    25.000000
35529    42.672428
35530    42.672428
35531    43.000000
35532    48.000000
35533    56.000000
35534    53.000000
35535    42.000000
35536    46.000000
35537    31.000000
35538    68.000000
35539    23.000000
35540    31.

We could also chose to create a subset of our data, only keeping rows that do not contain NaN values.

The point is to make conscious decisions about how to manage missing data. This is where we think about how our data will be used and how these values will impact the scientific conclusions made from the data.

Python gives us all of the tools that we need to account for these issues. We just need to be cautious about how the decisions that we make impact scientific results.

## Writing Out Data to CSV

We’ve learned about using manipulating data to get desired outputs. But we’ve also discussed keeping data that has been manipulated separate from our raw data. Something we might be interested in doing is working with only the columns that have full data. First, let’s reload the data so we’re not mixing up all of our previous manipulations.

In [45]:
surveys_df = pd.read_csv("data/surveys.csv")

Next, let’s drop all the rows that contain missing values. We will use the command dropna. By default, dropna removes rows that contain missing data for even just one column.

In [46]:
df_na = surveys_df.dropna()

If you now type df_na, you should observe that the resulting DataFrame has 30676 rows and 9 columns, much smaller than the 35549 row original.

We can now use the to_csv command to do export a DataFrame in CSV format. Note that the code below will by default save the data into the current working directory. We can save it to a different folder by adding the foldername and a slash before the filename: df.to_csv('foldername/out.csv'). We use ‘index=False’ so that pandas doesn’t include the index number for each line.

In [47]:
df_na.shape

(30676, 9)

In [49]:
# Write DataFrame to CSV
df_na.to_csv('data_output/surveys_complete.csv', index=False)